File: baekjoon/14500/14502.py
from itertools import combinations
from copy import deepcopy

N, M = map(int, input().split())

board = []
wall = []

for i in range(N):
    board.append(list(map(int, input().split())))
    for j in range(M):
        if board[i][j] == 0:
            wall.append((i,j))

def diffusion(board):
    for i in range(N):
        for j in range(M):
            if board[i][j] == 2:
                dfs(board, i, j)
    return

def dfs(board, ci, cj):
    di = [0,0,1,-1]
    dj = [1,-1,0,0]

    for d in range(4):
        ni, nj = ci+di[d], cj+dj[d]
        if not(0 <= ni < N and 0 <= nj < M): continue
        if board[ni][nj] != 0 : continue
        board[ni][nj] = -1
        dfs(board, ni, nj)




    return

def count(board):
    cnt = 0
    for i in range(N):
        for j in range(M):
            if board[i][j] == 0:
                cnt +=1
    return cnt

res = 0

for c in combinations(wall, 3):
    # 벽 설치
    for x,y in c:
        board[x][y] = 1

    temp = deepcopy(board)
    diffusion(temp)
    res = max(res, count(temp))

    # 벽 초기화
    for x,y in c:
        board[x][y] = 0

print(res)

File: baekjoon/14500/14502.java
/* package whatever; // don't place package name! */

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/* Name of the class has to be "Main" only if the class is public. */
public class Main {

  static class Pair {

    public int i;
    public int j;

    @Override
    public String toString() {
      return "Pair{" +
          "i=" + i +
          ", j=" + j +
          '}';
    }

    public Pair(int i, int j) {
      this.i = i;
      this.j = j;

    }
  }

  static int board[][];
  static int map[][];
  static int N;
  static int M;
  static List<Pair> wall;
  static List<Pair> combWall;
  static int ans;


  public static void main(String[] args) throws java.lang.Exception {
    Scanner s = new Scanner(System.in);

    N = s.nextInt();
    M = s.nextInt();
    board = new int[N][M];
    map = new int[N][M];
    wall = new ArrayList<>();
    combWall = new ArrayList<>();
    ans = 0;

    for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
        board[i][j] = s.nextInt();
        if (board[i][j] == 0) {
          wall.add(new Pair(i, j));
        }
      }
    }

    solve(-1, 0);
    System.out.println(ans);
  }

  public static void solve(int index, int size) {
    if (size == 3) {
      for (int i = 0; i < N; i++) for (int j = 0; j < M; j++) map[i][j] = board[i][j];
      combWall.forEach(w -> map[w.i][w.j] = 1);
      diffusion();
      ans = Math.max(ans, count());
      return;
    }

    for (int i = index + 1; i < wall.size(); i++) {
      combWall.add(wall.get(i));
      solve(i, size + 1);
      combWall.remove(combWall.size() - 1);
    }
  }

  public static void diffusion() {
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
        if(map[i][j] == 2) {
          dfs(i,j);
        }
      }
    }
  }

  public static void dfs(int ci, int cj) {
    int[] di = {0, 0, 1, -1};
    int[] dj = {1, -1, 0, 0};

    for (int d = 0; d < 4; d++) {
      int ni = ci + di[d];
      int nj = cj + dj[d];

      if (!((0 <= ni && ni < N) && (0 <= nj && nj < M))) continue;
      if (map[ni][nj] != 0) continue;
      map[ni][nj] = -1;
      dfs(ni,nj);
    }
  }

  public static int count() {
    int count = 0;
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
        if(map[i][j] == 0) count += 1;
      }
    }
    return count;
  }

}
